### PR TITLE
Fix 151

### LINKED ISSUE #151: Assertion failure in `decodeWitnessData`
`decodeWitnessData` calls `readBitstring`, which has an assertion on the length to read. But the length to read is an unchecked value read from the serialized program. So you can trigger an assertion.

`761f001a00`, whatever that is, seems to trigger this.

Found by fuzzer.

### PR DESCRIPTION
**Pull request by roconnor-blockstream**
Fixes #151.

### CODE REVIEW

**Comment by roconnor-blockstream** [comment]
@apoelstra Do you prefer this PR squashed?

**Comment by apoelstra** [comment]
Also, can you confirm that if somebody tries to make a witness of size 2^32 + 1, that the program is rejected without the assertion triggering? I believe the original bug occured in `decodeWitnessData` which doesn't understand types or anything, it just reads the total witness length and tries to cram it all into a bitstring.

**Comment by apoelstra** [comment]
Finally, I am getting build failures like
```
       > test.c:432:13: error: function declaration isn't a prototype [-Werror=strict-prototypes]
       >   432 | static void regression_tests() {
       >       |             ^~~~~~~~~~~~~~~~
       > cc1: all warnings being treated as errors
```
I suspect because I have a newer gcc than you. I think you have to stick a `void` into the argument list for the function.

**Comment by roconnor-blockstream** [comment]
Oops.  I was testing with clang.

**Comment by roconnor-blockstream** [comment]
Completed all suggestions.

File: C/test.c [hunk 1]
```diff
@@ -435,6 +435,20 @@ static void regression_tests(void) {
     const unsigned char regression1[] = {0x27, 0xe1, 0xe0, 0x00, 0x00, 0x00, 0x00};
     test_program("regression1", regression1, sizeof(regression1), SIMPLICITY_ERR_DATA_OUT_OF_RANGE, NULL, NULL, NULL);
   }
+  {
+    /* Unit program with incomplete witness of size 2^31-1. */
+    const unsigned char regression2[] = {0x27, 0xe1, 0xdf, 0xff, 0xff,  0xff, 0xff};
```
Comment (apoelstra) [inline C/test.c:440]: What if the incomplete witness has size 2^31 + 1?

**Review by apoelstra — APPROVED** [review]
ACK 8d2bdb426a55bff6c19ff8fd371d4abdd71f50b3

**Pull request merged**